winch/codegen/src/trampoline.rs

```diff
@@ -0,0 +1,176 @@
+use crate::{
+    abi::{align_to, calculate_frame_adjustment, ABIArg, ABIResult, ABI},
+    masm::{CalleeKind, MacroAssembler, OperandSize, RegImm},
+    reg::Reg,
+};
+use std::mem;
+use wasmparser::{FuncType, ValType};
+
+/// A trampoline to provide interopt between different calling conventions.
+pub(crate) struct Trampoline<'a, A, M>
+where
+    A: ABI,
+    M: MacroAssembler,
+{
+    /// The macro assembler.
+    masm: &'a mut M,
+    /// The ABI.
+    abi: &'a A,
+    /// The main scratch register for the current architecture. It is not allocatable for the callee.
+    scratch_reg: Reg,
+    /// A second scratch register. This will be allocatable for the callee, so it can only be used
+    /// after the callee-saved registers are on the stack.
+    alloc_scratch_reg: Reg,
+}
+
+impl<'a, A, M> Trampoline<'a, A, M>
+where
+    A: ABI,
+    M: MacroAssembler,
+{
+    /// Create a new trampoline.
+    pub fn new(masm: &'a mut M, abi: &'a A, scratch_reg: Reg, alloc_scratch_reg: Reg) -> Self {
+        Self {
+            masm,
+            abi,
+            scratch_reg,
+            alloc_scratch_reg,
+        }
+    }
+
+    /// Emit the host to wasm trampoline.
+    pub fn emit_host_to_wasm(&mut self, ty: &FuncType) {
+        // The host to wasm trampoline is currently hard coded (see vmcontext.rs in the
+        // wasmtime-runtime crate, VMTrampoline)
+        // The first two parameters are VMContexts (not used at this time)
+        // The third parameter is the function pointer to call
+        // The fourth parameter is an address to storage space for both the return value and the
+        // arguments to the function.
+        let trampoline_ty = FuncType::new(
+            vec![ValType::I64, ValType::I64, ValType::I64, ValType::I64],
+            vec![],
+        );
+
+        // TODO: We should be passing a calling convention here so the signature can determine the
+        // correct location of arguments. When we fully support system ABIs, this will need to be
+        // updated.
+        let trampoline_sig = self.abi.sig(&trampoline_ty);
+
+        // Hard-coding the size in bytes of the trampoline arguments since it's static, based on
+        // the current signature we should always have 4 arguments, each of which is 8 bytes
```

Comment:
```suggestion
        // Hard-coding the size in bytes of the trampoline arguments since it's static, based on
        // the current signature we should always have 4 arguments, each of which is 8 bytes.